### DISCUSSION TITLE
/digging-deeper/queues.html

### DISCUSSION THREAD

**Discussion by giscus** [goravel/goravel#125] [Announcements]
# /digging-deeper/queues.html

[[toc]] Introduction While building your web application, you may have some tasks, such as parsing and storing an uploaded CSV file, that take too long to perform during a web r...

https://www.goravel.dev/digging-deeper/queues.html

**Comment by ocano-diego** [comment]
do queues implement goroutines and channels?

**Reply by hwbrzzl** [reply]
goroutines and channels? Can you be more specific, please?

**Reply by ocano-diego** [reply]
with laravel you need a queue worker because PHP does not have good support for concurrency, with go you can use a goroutine to execute side jobs just wondering if the queue implementes this or you need to run an artisan command like queue:work like you do on laravel

**Reply by hwbrzzl** [reply]
Yes, we need to use goroutine to start the queue server, as shown below([the documentation](https://www.goravel.dev/digging-deeper/queues.html#start-queue-server)),  we implement the queue module by a third package: github.com/RichardKnop/machinery/v2.

<img width="753" alt="image" src="https://github.com/goravel/goravel/assets/24771476/33646e4b-75cd-4b77-a03c-a13cd5363fdd">

**Comment by diaojinlong** [comment]
Failed queues cannot be saved and retried

**Reply by hwbrzzl** [reply]
Yes, we have a plan to add these features, please look into: https://github.com/goravel/goravel/issues/153

**Comment by truongnhan0311** [comment]
how to retry Failed Job ?

**Reply by hwbrzzl** [reply]
Still in progress

**Comment by darakanoit** [comment]
Is there any problem if I have more than one instance of the application running, for example, for scaling in kubernetes?(race status or duplicate jobs?) In laravel, supervisor works separately from the main application

**Reply by hwbrzzl** [reply]
You can use this feature in multiple Pods, because each pod shares the same Redis server that stores the queues.

**Comment by ruoxiao-zh** [comment]
When will support be provided for the Kafka message queue?

**Reply by hwbrzzl** [reply]
We are refactoring the queue facade: https://github.com/goravel/goravel/issues/153, the Kafka driver will be implemented easily after that.

**Comment by tekpriest** [comment]
I got an error here
```bash
[2024-11-17 05:57:27] local.error: Queue run error: job signature duplicate: send_email_notification, the names of Job and Listener cannot be duplicated
trace:
        log.(*Writer).withStackTrace:/Users/tp/go/pkg/mod/github.com/goravel/framework@v1.14.7/log/logrus_writer.go:200
        log.(*Writer).Errorf:/Users/tp/go/pkg/mod/github.com/goravel/framework@v1.14.7/log/logrus_writer.go:103
```
This is the main.go file
```go
func main() {
	// This bootstraps the framework and gets it ready for use.
	bootstrap.Boot()

	// Create a channel to listen for OS signals
	quit := make(chan os.Signal, 1)
	signal.Notify(quit, syscall.SIGINT, syscall.SIGTERM)

	// Start http server by facades.Route().
	go func() {
		if err := facades.Route().Run(); err != nil {
			facades.Log().Errorf("Route Run error: %v", err)
		}
	}()

	// Listen for the OS signal
	go func() {
		<-quit
		if err := facades.Route().Shutdown(); err != nil {
			facades.Log().Errorf("Route Shutdown error: %v", err)
		}

		os.Exit(0)
	}()

	go func() {
		if err := facades.Queue().Worker(nil).Run(); err != nil {
			facades.Log().Errorf("Queue run error: %v", err)
		}
	}()

	select {}
}
```
and this is the send_email_notification.go file
```go
type SendEmailNotification struct {
	Z service_providers.Zeptomail
}

func (receiver *SendEmailNotification) Signature() string {
	return "send_email_notification"
}

func (receiver *SendEmailNotification) Queue(args ...any) event.Queue {
	return event.Queue{
		Enable:     true,
		Connection: "",
		Queue:      "",
	}
}

func (r *SendEmailNotification) Handle(args ...any) error {
	...
	return nil
}
```

**Reply by tekpriest** [reply]
setting event.Queue{Enable: false} still causes same error

**Reply by tekpriest** [reply]
seen the issue, this happens when a listener is added to more than one events in the event_service_provider file, is this by design?

**Reply by krishankumar01** [reply]
The framework uses a signature to register jobs and listeners. If you have two with the same signature, it will throw a duplicate error. Please ensure each signature is unique.

**Reply by hwbrzzl** [reply]
@tekpriest How do you register them now?

**Reply by tekpriest** [reply]
@hwbrzzl like this
```go
package providers

import (
	...

	"github.com/goravel/framework/contracts/event"
	"github.com/goravel/framework/contracts/foundation"
	"github.com/goravel/framework/facades"

..
)

type EventServiceProvider struct{}

func (receiver *EventServiceProvider) Register(app foundation.Application) {
	facades.Event().Register(receiver.listen())
}

func (receiver *EventServiceProvider) Boot(app foundation.Application) {}

func (receiver *EventServiceProvider) listen() map[event.Event][]event.Listener {
	return map[event.Event][]event.Listener{
		...
		&events.MemberRemoved{}: {
			&listeners.SendEmailNotification{
				Z:         *service_providers.NewZeptomailService(),
				QueueName: "member_removed",
			},
		},
		&events.MemberCreated{}: {
			&listeners.SendEmailNotification{
				Z:         *service_providers.NewZeptomailService(),
				QueueName: "member_created",
			},
		},
	}
}
```
and in the listener, I go wit this
```go
...
type SendEmailNotification struct {
	Z         service_providers.Zeptomail
	QueueName string
}

func (receiver *SendEmailNotification) Signature() string {
	return receiver.QueueName
}
...
```

**Reply by hwbrzzl** [reply]
Thanks for your feedback, and yes, it's a bug, I'll make a quick fix and release it in v1.14.8.

**Reply by hwbrzzl** [reply]
Hey @tekpriest This has been solved, please upgrade framework to v1.14.8, thanks for your feedback!

**Reply by tekpriest** [reply]
keep up the good work

**Comment by jxgame** [comment]
目前框架支持两种队列驱动： redis 和 sync。

如果使用 sync 驱动，它是即时执行的，和一个普通函数没有区别，是吗？
同时Delay 方法也是无效的，对不对？

**Reply by hwbrzzl** [reply]
是的，Delay 将无效。

**Reply by jxgame** [reply]
感谢，使用Redis之后 Delay 有用了。

**Comment by islamsamy214** [comment]
I think this might be useful, since u r using Gin, https://gin-gonic.com/docs/examples/goroutines-inside-a-middleware

Its just perfect and simple use of goroutine to run a background task, and its better than github.com/RichardKnop/machinery I think, of course its not the best for the queues, but we might make a benefit or make a queue with it and without depending on a third party!, what do u think?

**Reply by hwbrzzl** [reply]
Yes, there is [a PR](https://github.com/goravel/framework/pull/314) to optimize the queue module.
